File: main/store/persistence.ts
/**
 * JSON 文件持久化
 *
 * - 每个实体类型一个 JSON 文件
 * - 写入 500ms 防抖，原子写入（先 .tmp 再 rename）
 * - 启动时加载填充内存
 */

import * as fs from 'fs'
import * as path from 'path'
import type { AgentDefinition, Group, GroupMember, Task, Message, GroupFile, GroupQueueSnapshot } from './types'

const DATA_DIR = path.join(process.cwd(), 'data')

// 防抖定时器
const _debounceTimers = new Map<string, NodeJS.Timeout>()

// 队列防抖定时器（按 groupId）
const _queueDebounceTimers = new Map<string, NodeJS.Timeout>()

// 确保数据目录存在
function ensureDataDir(): void {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true })
  }
  const groupFilesDir = path.join(DATA_DIR, 'group_files')
  if (!fs.existsSync(groupFilesDir)) {
    fs.mkdirSync(groupFilesDir, { recursive: true })
  }
}

/**
 * 原子写入 JSON 文件
 * 先写 .tmp 再 rename，防止崩溃时数据丢失
 */
function atomicWrite(filePath: string, data: unknown): void {
  const tmpPath = filePath + '.tmp'
  const json = JSON.stringify(data, null, 2)
  fs.writeFileSync(tmpPath, json, 'utf-8')
  fs.renameSync(tmpPath, filePath)
}

/**
 * 读取 JSON 文件，文件不存在或损坏时返回空数组
 */
function readJsonFile<T>(filePath: string): T[] {
  try {
    if (!fs.existsSync(filePath)) return []
    const content = fs.readFileSync(filePath, 'utf-8')
    return JSON.parse(content) as T[]
  } catch {
    return []
  }
}

/**
 * 防抖写入：500ms 内多次写入只执行最后一次
 */
export function scheduleSave(entityName: string, data: unknown): void {
  const existing = _debounceTimers.get(entityName)
  if (existing) clearTimeout(existing)

  const timer = setTimeout(() => {
    ensureDataDir()
    const filePath = path.join(DATA_DIR, `${entityName}.json`)
    try {
      atomicWrite(filePath, data)
    } catch (err) {
      console.error(`Failed to save ${entityName}:`, err)
    }
    _debounceTimers.delete(entityName)
  }, 500)

  _debounceTimers.set(entityName, timer)
}

/**
 * 队列持久化：按 groupId 存储到 data/queues/{groupId}.json
 */
export function scheduleSaveQueue(groupId: string, data: GroupQueueSnapshot): void {
  const existing = _queueDebounceTimers.get(groupId)
  if (existing) clearTimeout(existing)

  const timer = setTimeout(() => {
    const queueDir = path.join(DATA_DIR, 'queues')
    if (!fs.existsSync(queueDir)) {
      fs.mkdirSync(queueDir, { recursive: true })
    }
    const filePath = path.join(queueDir, `${groupId}.json`)
    try {
      atomicWrite(filePath, data)
    } catch (err) {
      console.error(`Failed to save queue for group ${groupId}:`, err)
    }
    _queueDebounceTimers.delete(groupId)
  }, 500)

  _queueDebounceTimers.set(groupId, timer)
}

export async function loadAllQueues(): Promise<GroupQueueSnapshot[]> {
  const queueDir = path.join(DATA_DIR, 'queues')
  if (!fs.existsSync(queueDir)) return []

  const snapshots: GroupQueueSnapshot[] = []
  try {
    const files = fs.readdirSync(queueDir).filter(f => f.endsWith('.json'))
    for (const file of files) {
      try {
        const content = fs.readFileSync(path.join(queueDir, file), 'utf-8')
        snapshots.push(JSON.parse(content) as GroupQueueSnapshot)
      } catch {
        // 忽略损坏的队列文件
      }
    }
  } catch {
    return []
  }
  return snapshots
}

/**
 * 立即刷盘（应用退出时调用）
 */
export async function flushPersistence(): Promise<void> {
  // 清除所有实体防抖定时器
  for (const [, timer] of _debounceTimers) {
    clearTimeout(timer)
  }
  _debounceTimers.clear()

  // 清除所有队列防抖定时器
  for (const [, timer] of _queueDebounceTimers) {
    clearTimeout(timer)
  }
  _queueDebounceTimers.clear()
}

/**
 * 启动时加载所有数据
 */
export async function initPersistence(): Promise<void> {
  ensureDataDir()
}

export async function loadAll(): Promise<{
  agents: AgentDefinition[]
  groups: Group[]
  members: GroupMember[]
  tasks: Task[]
  messages: Message[]
}> {
  ensureDataDir()
  return {
    agents: readJsonFile<AgentDefinition>(path.join(DATA_DIR, 'agents.json')),
    groups: readJsonFile<Group>(path.join(DATA_DIR, 'groups.json')),
    members: readJsonFile<GroupMember>(path.join(DATA_DIR, 'members.json')),
    tasks: readJsonFile<Task>(path.join(DATA_DIR, 'tasks.json')),
    messages: readJsonFile<Message>(path.join(DATA_DIR, 'messages.json')),
  }
}

/**
 * 列出群组文件
 */
export function listFiles(groupId: string): GroupFile[] {
  const groupDir = path.join(DATA_DIR, 'group_files', groupId)
  if (!fs.existsSync(groupDir)) return []

  try {
    return fs.readdirSync(groupDir).map(name => {
      const filePath = path.join(groupDir, name)
      const stat = fs.statSync(filePath)
      return {
        name,
        size: stat.size,
        modified_at: stat.mtime.toISOString(),
      }
    })
  } catch {
    return []
  }
}

export const persistence = {
  scheduleSave,
  scheduleSaveQueue,
  flushPersistence,
  initPersistence,
  loadAll,
  loadAllQueues,
  listFiles,
}

File: src-tauri/src/core/persistence.rs
//! JSON 持久化 —— greenfield 重写
//!
//! 保留语义：每实体一个 JSON 文件（裸数组）+ queues/<group>.json 快照；
//! 写入 500ms 防抖 + 原子写（.tmp → rename）；加载时缺失/损坏返回空。
//! byte 兼容旧 data/*.json。

use crate::core::types::*;
use parking_lot::Mutex;
use serde::de::DeserializeOwned;
use std::path::{Path, PathBuf};
use std::sync::LazyLock;
use std::time::Duration;

static DEBOUNCE_TIMERS: LazyLock<Mutex<std::collections::HashMap<String, DebounceEntry>>> =
    LazyLock::new(|| Mutex::new(std::collections::HashMap::new()));

#[derive(Clone)]
struct DebounceEntry {
    handle_id: u64,
    payload: serde_json::Value,
    kind: DebounceKind,
}

#[derive(Clone, Copy, PartialEq)]
enum DebounceKind {
    Entity,
    Queue,
}

/// 全局自增 seq，标记最新一次防抖（旧的失效）
static DEBOUNCE_SEQ: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(0);

static RUNTIME_DATA_DIR: LazyLock<Mutex<PathBuf>> = LazyLock::new(|| Mutex::new(PathBuf::from(".")));

pub fn init(data_dir: &Path) {
    set_data_dir(data_dir);
    ensure_dirs(data_dir);
}

fn set_data_dir(p: &Path) {
    *RUNTIME_DATA_DIR.lock() = p.to_path_buf();
}

pub fn data_dir() -> PathBuf {
    RUNTIME_DATA_DIR.lock().clone()
}

fn ensure_dirs(base: &Path) {
    std::fs::create_dir_all(base).ok();
    std::fs::create_dir_all(base.join("group_files")).ok();
    std::fs::create_dir_all(base.join("queues")).ok();
    std::fs::create_dir_all(base.join("logs")).ok();
}

fn atomic_write(file_path: &Path, data: &serde_json::Value) -> std::io::Result<()> {
    let tmp = file_path.with_extension("json.tmp");
    let json = serde_json::to_string_pretty(data)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
    std::fs::write(&tmp, json)?;
    std::fs::rename(&tmp, file_path)?;
    Ok(())
}

fn read_json_file<T: DeserializeOwned>(file_path: &Path) -> Vec<T> {
    match std::fs::read_to_string(file_path) {
        Ok(content) => serde_json::from_str(&content).unwrap_or_default(),
        Err(_) => Vec::new(),
    }
}

fn read_queue_file(file_path: &Path) -> Option<GroupQueueSnapshot> {
    let content = std::fs::read_to_string(file_path).ok()?;
    serde_json::from_str(&content).ok()
}

pub fn load_all_queues() -> Vec<GroupQueueSnapshot> {
    let dir = data_dir().join("queues");
    let mut out = Vec::new();
    if let Ok(entries) = std::fs::read_dir(&dir) {
        for entry in entries.flatten() {
            let p = entry.path();
            if p.extension().and_then(|e| e.to_str()) == Some("json") {
                if let Some(snap) = read_queue_file(&p) {
                    out.push(snap);
                }
            }
        }
    }
    out
}

pub fn load_all() -> PersistedData {
    let base = data_dir();
    ensure_dirs(&base);
    PersistedData {
        agents: read_json_file(&base.join("agents.json")),
        groups: read_json_file(&base.join("groups.json")),
        members: read_json_file(&base.join("members.json")),
        tasks: read_json_file(&base.join("tasks.json")),
        messages: read_json_file(&base.join("messages.json")),
        queues: load_all_queues()
            .into_iter()
            .map(|q| (q.group_id.clone(), q))
            .collect(),
    }
}

fn schedule_save(kind: DebounceKind, key: String, data: serde_json::Value) {
    let seq = DEBOUNCE_SEQ.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
    {
        let mut timers = DEBOUNCE_TIMERS.lock();
        timers.insert(
            key.clone(),
            DebounceEntry {
                handle_id: seq,
                payload: data,
                kind,
            },
        );
    }
    let key_cloned = key;
    let _ = tauri::async_runtime::spawn(async move {
        tokio::time::sleep(Duration::from_millis(500)).await;
        flush_one(&key_cloned, seq);
    });
}

pub fn schedule_save_entity(entity_name: &str, data: serde_json::Value) {
    schedule_save(DebounceKind::Entity, entity_name.to_string(), data);
}

pub fn schedule_save_queue(group_id: &str, data: serde_json::Value) {
    schedule_save(DebounceKind::Queue, group_id.to_string(), data);
}

fn flush_one(key: &str, seq: u64) {
    let entry = {
        let mut timers = DEBOUNCE_TIMERS.lock();
        match timers.get(key).map(|e| (e.handle_id == seq, e.clone())) {
            Some((is_latest, entry)) if is_latest => {
                timers.remove(key);
                Some(entry)
            }
            _ => None,
        }
    };
    if let Some(entry) = entry {
        let base = data_dir();
        let path = match entry.kind {
            DebounceKind::Entity => base.join(format!("{key}.json")),
            DebounceKind::Queue => {
                let d = base.join("queues");
                std::fs::create_dir_all(&d).ok();
                d.join(format!("{key}.json"))
            }
        };
        if let Err(e) = atomic_write(&path, &entry.payload) {
            log::error!("[persistence] save {key} failed: {e}");
        }
    }
}

pub fn flush_all() {
    let entries: Vec<(String, DebounceEntry)> = DEBOUNCE_TIMERS.lock().drain().collect();
    let base = data_dir();
    for (key, entry) in entries {
        let path = match entry.kind {
            DebounceKind::Entity => base.join(format!("{key}.json")),
            DebounceKind::Queue => base.join("queues").join(format!("{key}.json")),
        };
        let _ = atomic_write(&path, &entry.payload);
    }
}

/// 列出群组工作目录下的文件（供 group_list_files 命令）
pub fn list_files(group_id: &str) -> Vec<GroupFile> {
    let dir = data_dir().join("group_files").join(group_id);
    let mut out = Vec::new();
    if let Ok(entries) = std::fs::read_dir(&dir) {
        for entry in entries.flatten() {
            let p = entry.path();
            if let Ok(meta) = entry.metadata() {
                if meta.is_file() {
                    let modified = meta
                        .modified()
                        .ok()
                        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                        .map(|d| {
                            chrono::DateTime::<chrono::Utc>::from_timestamp(d.as_secs() as i64, 0)
                                .map(|dt| dt.to_rfc3339())
                                .unwrap_or_default()
                        })
                        .unwrap_or_default();
                    out.push(GroupFile {
                        name: p
                            .file_name()
                            .and_then(|n| n.to_str())
                            .unwrap_or("")
                            .to_string(),
                        size: meta.len(),
                        modified_at: modified,
                    });
                }
            }
        }
    }
    out
}

/// 群组工作目录
pub fn group_work_dir(group_id: &str) -> PathBuf {
    data_dir().join("group_files").join(group_id)
}
